### ISSUE TITLE
Tracking of Fuser Issues for HuggingFace Bert Model

### ISSUE THREAD

**Issue by kevinstephano** [csarofeen/pytorch#521]
## 🐛 Bug

**Docker Run Command**
```
nvidia-docker run -it --rm --net=host --uts=host --ipc=host --ulimit stack=67108864 --ulimit memlock=-1 --cap-add=SYS_ADMIN --cap-add=SYS_PTRACE --security-opt seccomp=unconfined -v [PATH ON MACHINE TO HuggingFace 
 CODE]:/my_workspace/ gitlab-master.nvidia.com/dl/pytorch/update-scripts:pytorch-nightly-jit-[CURRENT Date like 02032021]-cuda11
```

**Setup APEX for Fused Optimizers in Container**
```
cd /opt/pytorch/pytorch
git clone https://github.com/NVIDIA/apex
cd apex
pip install -v --disable-pip-version-check --no-cache-dir --global-option="--cpp_ext" --global-option="--cuda_ext" ./
```
**HuggingFace Fork**
https://github.com/kevinstephano/transformers

Command to setup HuggingFace:
```
python setup.py develop && pip install -r ./examples/_tests_requirements.txt
```

Command to run perf from HuggingFace Fork:
```
cd [HuggingFace Repo]/examples/text-classification

# For TF32 on Ampere
python run_perf.py --fused_adam

# For FP32 you need to run a smaller batch size
python run_perf.py --batch_size=64 --fused_adam
```
* To profile: `--profile`
* Turn off jit: `--nojit`
* Print kernels: `--print-kernel`
* Run just 10 steps: `--ten_steps`
* Turn off dynamic batches: `--pad`
* Epochs: `--epochs=[number: default of 20.0]`
* Batch Size: `--batch_size=[number: default of 256]`
* Fused Adam: `--fused_adam`
* TensorExpr Fuser: `--te`  use the TensorExpr fuser instead of NVFuser

**Dropout**
This requires [Jie's Branch](https://github.com/csarofeen/pytorch/tree/kevin_fix_dropout)
It does two things:
* Fixes the problem with saving too many intermediate tensors
* Removes conditionals for implementation defined in `symbolic_script.cpp`.

**Gelu**
Need to add the following code to `symbolic_script.cpp`.  This is found in a [upstream PR](https://github.com/pytorch/pytorch/pull/46785)

```
        def gelu(self):
            result = torch.gelu(self)
            def backward(grad_output):
                return 0.5 * (1. + torch.erf(self * 0.70710678)) + 0.3989423 * self * torch.exp(-0.5 * self * self) * grad_output
            return result, backward
```

Also enabled pointwise fusion unrolling and set it to 4.  Unrolling was disabled because the Fusion contained randomization from dropout.

**Linear+Bias Modules**
The `nn.Linear` modules in pytorch have a conditional in their definition for eager mode fusion of a `bias` to a `matmul`.  This inhibits the fusion with jit-script and, currently, our workaround is to separate the `bias` from an `nn.Linear` definition of the `bias`.

**Bias Backward**
* `sum_to` support was added in [Shiming's PR](https://github.com/csarofeen/pytorch/pull/424) but it is not currently seen in a network. This requires `IValue` profiling that is found in [Jie's Upstream PR](https://github.com/pytorch/pytorch/pull/47668).
* This triggers the `add_0` function in `symbolic_script.cpp` that has a multiplication by an Alpha parameter.  This multiplication generates a spurious unneeded kernel that kills per.

```
        def add_0(self,
                  other,
                  *,
                  alpha: number):
            result = torch.add(self, other, alpha=alpha)
            self_size, other_size = AD_sizes_if_not_equal_multi_1(self, other, result)
            def backward(grad_output):
                #grad_other = (grad_output * alpha)._grad_sum_to_size(other_size)
                grad_other = (grad_output * alpha)._grad_sum_to_size(other_size)
                grad_self = (grad_output)._grad_sum_to_size(self_size)
                return grad_self, grad_other, None
            return result, backward
```

**Layer Norm**
There was an attempted layer norm implementation in `symbolic_script.cpp` but it is currently disabled due to a bug.  I think we should edit `symbolic_srcript.cpp` to look like Softmax and call the torch equivalent for Layer Norm.
 
We need a backward pass implemented for Layer Norm in `parser.cpp`.

**Softmax**

We need a backward pass defined for Layer Norm in `parser.cpp`.

**Comment by jjsjann123** [comment]
Related upstream issues:
https://github.com/pytorch/pytorch/issues/48089
https://github.com/pytorch/pytorch/issues/48090
https://github.com/pytorch/pytorch/issues/48091
https://github.com/pytorch/pytorch/issues/48094

**Comment by kevinstephano** [comment]
This is more documentation for HuggingFace Bert that is now out of date.
